src/jupyterlabcontroller/storage/k8s.py

```diff
@@ -674,3 +682,430 @@ def _standard_metadata(self, name: str) -> V1ObjectMeta:
+    in the FastAPI event loop, with no obvious way to connect the
+    stack trace back to the place you forgot an "await" or misspelled
+    a method name."""
+
+    # Methods for fileserver
+
+    async def create_fileserver(
+        self,
+        username: str,
+        namespace: str,
+        job: V1Job,
+        service: V1Service,
+        gf_ingress: dict[str, Any],
+    ) -> bool:
+        with FILESERVER_LOCK:
+            self._logger.info(f"Creating new fileserver for {username}...")
+            self._logger.debug(f"...creating new job for {username}")
+            await self._create_fileserver_job(username, namespace, job)
+            self._logger.debug(f"...creating new service for {username}")
+            await self._create_fileserver_service(
+                username, namespace, spec=service
+            )
+            self._logger.debug(f"...creating new gfingress for {username}")
+            await self._create_fileserver_gafaelfawringress(
+                username, namespace, spec=gf_ingress
+            )
+            self._logger.debug(
+                f"...polling until objects appear for {username}"
+            )
+            timeout = 60.0
+            interval = 3.9
+            try:
+                async with asyncio.timeout(timeout):
+                    while True:
+                        good = await self.check_fileserver_present(
+                            username, namespace
+                        )
+                        if good:
+                            self._logger.info(
+                                f"Fileserver created for {username}"
+                            )
+                            return True
+                        await asyncio.sleep(interval)
+            except asyncio.TimeoutError:
+                self._logger.error(
+                    f"Fileserver for {username} did not appear."
+                )
+                return False
+
+    async def check_namespace(self, namespace: str) -> bool:
+        """Check to see if namespace is present; return True if it is,
+        False if it is not."""
+        try:
+            await self.api.read_namespace(namespace)
+            return True
+        except ApiException as e:
+            if e.status != 404:
+                msg = f"Cannot read namespace {namespace}"
+                raise KubernetesError.from_exception(
+                    msg, e, namespace=namespace
+                ) from e
+            return False
+
+    async def _create_fileserver_job(
+        self, username: str, namespace: str, job: V1Job
+    ) -> None:
+        obj_name = f"{username}-fs"
+        self._logger.debug("Creating job", name=obj_name, namespace=namespace)
+        try:
+            await self.batch_api.create_namespaced_job(namespace, job)
+        except ApiException as e:
+            if e.status == 409:
+                # It already exists.  Delete and recreate it
```

Comment:
Could you explain more in comment / docstring why this is the right thing to do? I'm not sure I understand.

Reply:
The fileserver itself has no meaningful state.  If there are somehow leftover bits of a previous fileserver when we try to create a new one, we want to just destroy the previous pieces, and then continue with our object creation.

Reply:
Comments added.